keps/sig-node/5027-dra-admin-controlled-device-attributes/README.md

```diff
@@ -0,0 +1,868 @@
+If any of the resources can be exhausted, how this is mitigated with the existing limits
+(e.g. pods per node) or new limits added by this KEP?
+
+Are there any tests that were run/should be run to understand performance characteristics better
+and validate the declared limits?
+-->
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+Instead of standardized attributes, top-level fields could be introduced. But
+the ability to override all kinds of attributes may be useful by itself.  Once
+that is implemented, adding attributes with special semantic is simpler than
+dedicated fields.
+
+Instead of ResourceSliceOverride as a separate type, new fields in the
+ResourceSlice status could be modified by an admin. That has the problem that
+the ResourceSlice object might get deleted while doing cluster maintenance like
+a driver update, in which case the admin intent would get lost. A driver would
+not be able to publish a new ResourceSlice where a device is immediately marked
+as offline because creating a ResourceSlice strips the status.
```

Comment:
JFYI:

@johnbelamaric suggested drop-in file style to overide/extend device attributes [here](https://github.com/NVIDIA/k8s-dra-driver/issues/213#issuecomment-2590526324):
> I also think it could be useful for the driver (actually, the base driver framework that we would prefer all drivers to use) to have a hook to allow VM architects to augment the device attributes published by the driver.
> 
> For example, dropping a file on the node that can tell you which external network each NIC is plumbed to.
> 
> Patrick's KEP gives the cluster admin an opportunity to enhance attributes. That could be sufficient to do what I am saying. But it may also be helpful to have an on-node way of doing this.

Reply:
If DRA driver authors want to support a way of doing this, then they certainly can. But I don't think we as Kubernetes should standardize and require supporting such a feature. If we want to offer a common API, then this KEP looks like a better approach to me, in particular because accessing the apiserver is easier than creating files on nodes...

Reply:
> If we want to offer a common API, then this KEP looks like a better approach to me, in particular because accessing the apiserver is easier than creating files on nodes...

I also support the ResourceSliceOverride approach. Although drop-in might fit with node-local devices, DRA can now support broader device models, e.g. non-node-local devices (i.e. fabric-attached devices).